ateam_common/src/assignment.cpp

```diff
@@ -0,0 +1,288 @@
+// Copyright 2021 A Team
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "ateam_common/assignment.hpp"
+
+#include <unordered_set>
+#include <vector>
+
+#include <iostream>
+
+namespace ateam_common::assignment
+{
+std::unordered_map<std::size_t, std::size_t> optimize_assignment(
+  const Eigen::MatrixXd & cost_matrix)
+{
+  internal::CostMarkCovers cmc;
+  cmc.cost_matrix = internal::squarize_matrix(cost_matrix);
+
+  internal::apply_step_1(cmc);
+  internal::apply_step_2(cmc);
+  internal::apply_step_3(cmc);
+
+  do {
+    internal::apply_step_4(cmc);
+    internal::apply_step_5(cmc);
+  } while (!internal::has_unique_assignments(cmc));
+
+  std::unordered_map<std::size_t, std::size_t> assignment;
+  for (int i = 0; i < cmc.mark_matrix.rows(); i++) {
+    for (int j = 0; j < cmc.mark_matrix.cols(); j++) {
+      bool is_inbounds = i < cost_matrix.rows() && j < cost_matrix.cols();
```

Comment:
If we go out of bounds here, we can continue to the next row since the following columns will be out of bounds; although, it seems rare to me that this would occur (when there would be fewer possible assignments than total robots), so it may not be worth it to optimize

Reply:
Added quick continues on the row loop and the col loop instead of just the if statement